libnetwork/osl: restore the right thread's netns

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
- Fixed `panic: Log in goroutine after TestReplySERVFAIL/ProxyDNS=true has completed` flakes by no longer touching global logrus state
- Resolved the root cause of flakiness in `TestProxyNXDOMAIN` by fixing `osl.setIPv6` to capture and restore the locked-to thread's original network namespace instead of "restoring" it to the netns of the calling goroutine's thread
- Added a canary check to `TestProxyNXDOMAIN` which can diagnose this situation if it happens again

**- How to verify it**
Throw it at CI and see what sticks

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- Fixed a correctness issue in libnetwork which is believed to only impact tests

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
libnetwork: make resolver tests less confusing

tstwriter mocks the server-side connection between the resolver and the
container, not the resolver and the external DNS server, so returning
the external DNS server's address as w.LocalAddr() is technically
incorrect and misleading. Only the protocols need to match as the
resolver uses the client's choice of protocol to determine which
protocol to use when forwarding the query to the external DNS server.
While this change has no material impact on the tests, it makes the
tests slightly more comprehensible for the next person.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork: leave global logger alone in tests

Swapping out the global logger on the fly is causing tests to flake out
by logging to a test's log output after the test function has returned.
Refactor Resolver to use a dependency-injected logger and the resolver
unit tests to inject a private logger instance into the Resolver under
test.

Signed-off-by: Cory Snider <csnider@mirantis.com>

cpuguy83 commented:
```
[2023-05-19T16:24:28.437Z] === FAIL: libnetwork TestProxyNXDOMAIN (0.00s)
[2023-05-19T16:24:28.437Z]     resolver_test.go:522: Server socket is in network namespace NS(91: 4, 4026532540), but test goroutine is in NS(92: 4, 4026532686)
```

corhere committed:
libnetwork/osl: restore the right thread's netns

osl.setIPv6 mistakenly captured the calling goroutine's thread's network
namespace instead of the network namespace of the thread getting its
namespace temporarily changed. As this function appears to only be
called from contexts in the process's initial network namespace, this
mistake would be of little consequence at runtime. The libnetwork unit
tests, on the other hand, unshare network namespaces so as not to
interfere with each other or the host's network namespace. But due to
this bug, the isolation backfires and the network namespace of
goroutines used by a test which are expected to be in the initial
network namespace can randomly become the isolated network namespace of
some other test. Symptoms include a loopback network server running in
one goroutine being inexplicably and randomly being unreachable by a
client in another goroutine.

Capture the original network namespace of the thread from the thread to
be tampered with, after locking the goroutine to the thread.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
libnetwork: check for netns leaks from prior tests

TestProxyNXDOMAIN has proven to be susceptible to failing as a
consequence of unlocked threads being set to the wrong network
namespace. As the failure mode looks a lot like a bug in the test
itself, it seems prudent to add a check for mismatched namespaces to the
test so we will know for next time that the root cause lies elsewhere.

Signed-off-by: Cory Snider <csnider@mirantis.com>

cpuguy83 reviewed (approved):
LGTM assuming CI is happy
